### ISSUE TITLE
Only import `json` and `argparse`, if needed

### ISSUE THREAD

**Issue by sphh** [hoedt/upsilonconf#4]
In my case I use YAML files for configuration. Would it be possible to reorganize the code, so that neither `json` nor `argparser` are imported, if they are not used?

Since both modules are include in the standard Python, these imports should be ok. But they take time and use up memory, so it would be preferable to only import them, if needed!

**Comment by hoedt** [comment]
Technically, it should be possible to reorganise the code. However, I am not sure if this would break (some) type-checkers, since the `argparse` module is also used for type-hinting.

Did you observe high memory usage due to these imports? I have no experience with memory profiling in python, so it might be helpful if you could provide some insights as to why you think memory consumption is too high.

**Comment by sphh** [comment]
Have I already mentioned, that I do not use (and care about) type-checkers? :grin:

I don't know either, how to profile the memory consumption of imported modules but I found https://stackoverflow.com/questions/67157553/why-does-my-package-take-up-so-much-memory#answer-67205462 dealing with it …

This issue was more born out of a purely puristic approach: Why importing modules, which are not needed?

**Comment by hoedt** [comment]
I have looked into the issue a bit and it seems that this is a bit of a general python issue (see e.g. https://scientific-python.org/specs/spec-0001). I used some of the ideas to implement some form of lazy loading (in the `lazy_import` branch), but I have this feeling that it introduces quite a bit of overhead and I'm not sure if it is really worth it.

Would you be able to quantify the memory and/or runtime overhead of eager vs lazy loading in some way?

**Comment by sphh** [comment]
I wonder, why not putting the `argparse` stuff into a submodule, which is not automatically imported:
```python3
>>> # Import standard functionality ...
>>> import upsilonconf
>>> # ... and use it
>>> conf = upsilonconf.Configuraiion(...)
>>> # Using argparse stuff fails
>>> conf = upsilonconf.from_cli()
AttributeError
>>> conf = upsilonconf.argparse.from_cli()
AttributeError

>>> # Import argparse stuff ...
>>> import upsilonconf.argparse
>>> # ... and use it
>>> conf = upsilonconf.argparse.from_cli()
```

As for the loading and saving, my approach would have been to put these also into submodules:
```python3
import upsilonconf
import upsilonconf.json
import upsilonconf.yaml

conf = upsilonconf.load('config.json', loader=upsilonconf.json.JsonLoader)
upsilonconf.save(conf, 'config.json', dumper=upsilonconf.yaml.YamlDumper)
```
This would also have the added benefit, that it is easy to extend by users, who can write their own loaders/dumpers for their own file formats. The only disadvantage I can see, is that the user has to write the logic him/herself, to select the loader/dumper based on the file extension:
```python3
suffix = pathlib.Path(configfile).suffix
if suffix == '.json':
    loader = JsonLoader
elif suffix == '.yaml':
    loader = YamlLoader
elif suffix == '.my_suffix':
    loader = MyLoader
config = upsilonconf.load('configfile, loader=loader)
```
but to be honest, how often do you support different configuration formats in one application?

Just my 2¢ thoughts …

**Comment by hoedt** [comment]
The `json` and `yaml` packages are now all lazily loaded. So I would argue that the only "problem" would be the import of `argparse` (and/or `re` if you want).

Generally, the `save` and `load` functions are mainly there for convenience.
Normally, I would agree that a more flexible interface is preferred, but I am not sure this is the case here.
If you are already writing a custom loader, the easiest thing would probably be to write something like
(especially with the newly added features :wink: )

```python
with open(filename, "r") as fp:
    config = Configuration.from_dict(loader.load(fp))
```

The same holds for argument parsing: I just wanted a convenient tool to override the config via the command-line arguments.
I personally find it cumbersome if I have to use multiple import statements to get access to the tools I need.
Especially for a small library like this one. That is why I would prefer to leave everything accessible in the package.

PS: if you need more flexible IO operations, it would probably be better to start a new issue (since I would consider this a separate feature)

**Comment by sphh** [comment]
I took the liberty, to implement it, how I would have done it: https://github.com/sphh/upsilonconf/tree/loadersdumpers:
```python3
import upsilonconf
from upsilonconf.yaml import YAMLLoader
from upsilonconf.json import JSONLoader

# Load a yaml file
conf = upsilonconf.load('~/.config/upsilon/conf.yaml', YAMLLoader())

# Save a json file in our preferred format
dumper = JSONDumper(indent=3, sort_keys=True)
# Save as often as you want
upsilonconf.save(config, '~/.config/upsilon/conf.json', dumper)
...
upsilonconf.save(config, '~/.config/upsilon/conf.json', dumper)
```
Or for `cli`:
```python3
import upsilonconf
from upsilonconf.cli import CLI
from upsilonconf.json import JSONLoader
from upsilonconf.yaml import YAMLLoader

json_loader = JSONLoader()
yaml_loader = YAMLLoader()
loaders = {
    '.json': json_loader,
    '.jsn': json_loader,
    '.yaml': yaml_loader,
    '.yml': yaml_loader,
    '.cfg': yaml_loader
}
cli = CLI(loaders=loaders)
config = cli.from_cli(sys.argv[1:])
```

Advantages:
- No complicated lazy loading.
- Easy to extend (write your own loader or dumper).
- No hard-coded relationships between file suffixes and loaders/dumpers (Unix/Linux users are more accustomed to this concept than Windows user). This also give the user full control over the format, which is actually written.
- Clear API very similar to many other configuration libraries (e.g. `pyyaml`).

Personally I would also get rid of the `load()` and `save()` functions by implementing them into the `Configuration()` class as method and factory:
```python3
class Configuration(MutableMapping[str, Any]):
    def save(self, path: Union[Path, str], dumper, key_modifiers: Mapping[str, str] = {}) -> None:
        path = Path(path).expanduser().resolve()
        path.parent.mkdir(exist_ok=True, parents=True)
        config = _replace_in_keys(config, key_modifiers)
        dumper.save(config, path)

    @classmethod
    def load(cls, path: Union[Path, str], loader, key_modifiers: Mapping[str, str] = {}) -> Configuration:
        path = Path(path).expanduser().resolve()
        config = _replace_in_keys(loader.load(path), key_modifiers)
        return cls(**config)
```
Maybe the `dumper` and the `key_modifiers` used in saving could be stored inside the class … Also the `path`, which is used during loading, could be stored and used for the `save()` method, if no path is supplied?

Then you could use:
```python3
import upsilonconf
from upsilonconf.json import JSONLoader, JSONDumper

config = Configuration.load('~/.config/upsilon/config.cfg', JSONLoader())
dumper = JSONDumper(sort_keys=True)
...
config.save('~/.config/upsilon/config.cfg', dumper)
# or with the changes mentioned above, simply
config.save()
```

But …
- it breaks the existing API (obviously I believe, it is worthwhile and not too complicated to update).
- it is your code and your preferences about it!

**Comment by sphh** [comment]
PS: There exists now a new branch https://github.com/sphh/upsilonconf/tree/factory as proof of concept, which has the `load()` and `save()` functions moved into the `Configuration` class …

**Comment by hoedt** [comment]
I totally agree on the refactoring (see `io_refactoring` branch). I do not agree on putting IO logic in `Configuration` because the IO is really supposed to be something _extra_ (in my eyes).

However, the actual problem is that you would prefer to have the `io` stuff as an additional import that is not loaded by default, such that

```python
import upsilonconf.io

json_io =  upsilonconf.io.JSONIO()
config = json_io.load_config("/path/to/config.json")
yaml_io = upsilonconf.io.YAMLIO()
yaml_io.save_config(config, "/path/to/config.yaml")
```

or using my `FlexibleIO` conception:

```python
import upsilonconf.io

config_io =  upsilonconf.io.FlexibleIO({
    ".json": upsilonconf.io.JSONIO(),
    ".yaml": upsilonconf.io.YAMLIO()
})
config =config_io.load_config("/path/to/config.json")
config_io.save_config(config, "/path/to/config.yaml")
```

whereas I would prefer to have everything accessible directly from `upsilonconf`, such that

```python
import upsilonconf

config = upsilonconf.load_config("path/to/config.json")
upsilonconf.save_config(config, "/path/to/config.yaml")
```

I admit that the difference/additional complexity due to your approach is not as bad as I had it in mind.
It does require users to dig up all options and makes it harder/impossible to take the "I don't care"-route.
On the other hand, the benefit of the (small) increase in complexity is not really that pronounced either, is it?
I think I would really prefer to preserve the convenience of the original API.
By means of the refactoring, everybody should be able to implement more advanced scenarios if they want.

The only remaining problem would be the module loading overhead (without lazy imports).
I personally think this falls in the category of premature optimisation (this is not `numpy` or `scipy`).
However, your mileage may vary.

**Comment by sphh** [comment]
It is your project and your decision.

Having said that, I believe that a factory method to generate a class instance is becoming more and more popular.

Adding a `save` method is then a logical step (for me). And because of the `IO` classes, it is a very simple method. (Maybe the `IO` dumper should only return the string, which is then written to the file by the `Configuration`. The same with the loading: `Configuration.load()` sends the read data to the `IO` loader, which returns a dictionary.)

But the real benefit of the `IO` class approach is, that it is transparent (does not rely on obscure hard-coded file suffixes determining the `IO` format) and finally lends itself to supporting other formats easily. Finally no lazy loading tricks are required.

`pyyaml` takes the very same route, but even `json` supports it, so it shouldn't be completely new to developers.


I also believe, that `upsilonconf` target audience are developers and not users (in which case a  ‘I don't care’ approach would be the best route).

But your approach is ok as well! Nothing wrong with it. (I would just prefer to not load modules, which are not needed.)

**Comment by hoedt** [comment]
All of the convenience shortcuts actually use lazy loading now, so it seems that we both got what we wanted. Therefore, I think we can consider this issue resolved with the refactoring (commit 66140d5).

**Comment by sphh** [comment]
Glad, that a solution has been found!
